File: solutions/0624. Maximum Distance in Arrays/0624.py
# Time complexity: O(n)
# Space complexity: O(1)
class Solution:
  def maxDistance(self, arrays: List[List[int]]) -> int:
    ans = 0
    mn = 10000
    mx = -10000

    for arr in arrays:
      ans = max(ans, arr[-1] - mn, mx - arr[0])
      mn = min(mn, arr[0])
      mx = max(mx, arr[-1])

    return ans

File: solutions/0624. Maximum Distance in Arrays/0624.cpp
// Time complexity: O(n)
// Space complexity: O(1)
class Solution {
 public:
  int maxDistance(const vector<vector<int>>& arrays) {
    int ans = 0;
    int mn = 10000;
    int mx = -10000;

    for (const vector<int>& arr : arrays) {
      ans = max({ans, arr.back() - mn, mx - arr.front()});
      mn = min(mn, arr.front());
      mx = max(mx, arr.back());
    }

    return ans;
  }
};